### DISCUSSION TITLE
orm.em.execute(...) does not pickup the current context.

### DISCUSSION THREAD

**Discussion by nicomouss** [mikro-orm/mikro-orm#3896] [Q&A] [answered]
**Describe the bug**
`orm.em.execute(...)` is not executed in the current transactional context.

```js
@Injectable()
export class MyService {
  constructor(
    private readonly orm: MikroORM, 
    private readonly myRepository: MyRepository) { }

  // Wrapping the call to doSomething() in a RequestContext
  @UseRequestContext()
  async doSomething() {
    try {
      // Starting transaction
      await this.orm.em.begin();
      // Executing some stuff within the transaction...
      await this.myRepository.executeCustomSql();
      // Committing the transaction
      await this.orm.em.commit();
    } catch (error) {
      // Rolling back the transaction
      await this.orm.em.rollback();
      throw error;
    }
  }
}
```
```js
@Injectable()
export class MyRepository {
  constructor(private readonly orm: MikroORM) { }

  async executeCustomSql() {
    // Executing custom SQL instructions
    await (this.orm.em as SqlEntityManager).execute('create table .....');

    // Throwing error to force a transaction ROLLBACK in the wrapping service
    // => The custom SQL should be rolled back, but it is not...
    throw new Error('Error !!!');
  }
}
```

When looking at the logs, it seems that the SQL is indeed rolled back, but that is misleading information, as the SQL is actually not rolled back at all:
```
[query] begin
[query] create table ..... [took 10 ms]
[query] rollback
```

**Expected behavior**
`orm.em.execute(...)` should pickup the current RequestContext and transactional context, if any.

**Additional context**
Workaround is to access the `EntityManager` through the `RequestContext` directly:
```js
await (RequestContext.getEntityManager() as SqlEntityManager).execute('create table .....');
```
In that case, the current transactional context is properly respected, and things work fine.

**Versions**

| Dependency | Version |
| - | - |
| mikro-orm | 5.6.1 |
| your-driver | postgresql |

**Comment by B4nan** [comment, marked as answer]
I got confused by your repro, as you are missing many `await`s in there (which would end up with the same effect), but I see you are right, the `em.execute()` is indeed not using the right transaction context on its own.

> Workaround is to access the EntityManager through the RequestContext directly:

Much simpler workaround:

```ts
await this.em.getContext().execute('create table .....');
```

Anyway, fixed it [here](https://github.com/mikro-orm/mikro-orm/commit/832105d23de63df29010ccf62f4ec7a67955a47f), so you can use the dev version in few minutes too.

**Reply by nicomouss** [reply]
> I got confused by your repro, as you are missing many `await`s in there (which would end up with the same effect)

Sorry, my bad, indeed my `await` instructions went away as I cleaned up my code to make it more legible as a post here.
Thanks for pointing that out, I edited my repro code to make it the way I intended it to be.

> Much simpler workaround:
> 
> ```ts
> await this.em.getContext().execute('create table .....');
> ```
Thank you for the tip on using `em.getContext()` as another workaround.

> Anyway, fixed it [here](https://github.com/mikro-orm/mikro-orm/commit/832105d23de63df29010ccf62f4ec7a67955a47f), so you can use the dev version in few minutes too.

Thanks!
